File: morning_exercises/day10.cpp
#include <algorithm>
#include <cstdlib>
#include <ios>
#include <iostream>
#include <memory>
class TreeNode {
    public:
        int val;
        std::shared_ptr<TreeNode> left;
        std::shared_ptr<TreeNode> right;

    TreeNode(int val = 0): val(val), left(nullptr), right(nullptr) {}
};

int maxDepth(std::shared_ptr<TreeNode> root) {
    if (root==nullptr) return 0;

    return 1 + std::max(maxDepth(root->left), maxDepth(root->right));
}

int isBalancedHelper(std::shared_ptr<TreeNode> root) {
    if (root == nullptr) return 0;
    int left = isBalancedHelper(root->left);
    int right = isBalancedHelper(root->right);

    if (left==-1) return -1;
    if (right==-1) return -1;

    if (std::abs(left-right)>1) return -1;
    return 1 + std::max(left, right);
}

bool isBalanced(std::shared_ptr<TreeNode> root) {
    return isBalancedHelper(root)!=-1;
}

int main() {
    auto root = std::make_shared<TreeNode>(15);
    root->left = std::make_shared<TreeNode>(14);
    root->right = std::make_shared<TreeNode>(16);
    root->left->left = std::make_shared<TreeNode>(12);
    root->left->right = std::make_shared<TreeNode>(14);
    root->left->left->left = std::make_shared<TreeNode>(11);
    std::cout << std::boolalpha;
    std::cout << isBalanced(root) << "\n";
}

File: morning_exercises/day10.py
class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

def isBalancedHelper(root: TreeNode | None)->int:
    if (root is None): return 0

    left = isBalancedHelper(root.left)
    right = isBalancedHelper(root.right)

    if (left==-1): return -1
    if (right==-1): return -1

    if (abs(left-right)>1): return -1
    return 1+max(left, right)

def isBalanced(root: TreeNode | None)->bool:
    return isBalancedHelper(root)!=-1

root = TreeNode(12)
root.left = TreeNode(4)
root.right = TreeNode(7)
root.right.left = TreeNode(5)
root.right.right = TreeNode(8)
# root.right.right.right = TreeNode(10)

print(isBalanced(root))